File: Sentence.cpp
#include <iostream>
#include <vector>
#include <map>
#include <set>

bool isSentence(std::vector<std::string>&);

int main()
{
	std::vector<std::string> one = { "the", "dog", "ran", "EOI" };
	std::vector<std::string> two = { "the", "lazy", "dog", "ran", "slowly", "EOI" };
	std::vector<std::string> three = { "dog", "ran", "EOI" };
	std::vector<std::string> four = { "the", "lazy", "lazy", "lazy", "smelly", "dog", "ran", "EOI" };
	std::vector<std::string> five = { "the", "dog", "ran", "noisily" };

	std::vector<std::vector<std::string>> list;
	list.push_back(one);
	list.push_back(two);
	list.push_back(three);
	list.push_back(four);
	list.push_back(five);

	for (int i = 0; i < list.size(); i++)
	{
		bool valid = isSentence(list.at(i));
		std::cout << "Sentence " << i << ": ";
		for (int j = 0; j < list.at(i).size(); j++)
		{
			std::cout << list.at(i).at(j) << " ";
		}
		std::cout << "- Valid: ";
		if (valid == true)
		{
			std::cout << "true" << std::endl;
		}
		else
		{
			std::cout << "false" << std::endl;
		}
	}
	return 0;
}

bool isSentence(std::vector<std::string>& sentence)
{
	std::map<std::pair<std::string, std::string>, std::string> table {{std::make_pair("S0", "the"),    "S1"},
																	  {std::make_pair("S1", "lazy"),   "S2"},  {std::make_pair("S1", "smelly"),  "S2"},
																	  {std::make_pair("S1", "dog"),    "S3"},  {std::make_pair("S1", "cat"),     "S3"},
																	  {std::make_pair("S2", "lazy"),   "S2"},  {std::make_pair("S2", "smelly"),  "S2"},
																	  {std::make_pair("S2", "dog"),    "S3"},  {std::make_pair("S2", "cat"),     "S3"},
																	  {std::make_pair("S3", "ate"),    "S4"},  {std::make_pair("S3", "ran"),     "S4"},
																	  {std::make_pair("S4", "slowly"), "S5"},  {std::make_pair("S4", "noisily"), "S5"},
																	  {std::make_pair("S4", "EOI"),    "END"}, {std::make_pair("S5", "EOI"),     "END"}};

	std::pair<std::string, std::string> temp;
	std::string currState = "S0";
	for (int i = 0; i < sentence.size(); i++)
	{
		temp = std::make_pair(currState, sentence.at(i));
		if (table.count(temp) == 1)
		{
			currState = table.at(temp);
			if (currState == "END")
			{
				return true;
			}
		}
		else
		{
			return false;
		}
	}
}

File: sentence.js
const states_tab = {
    0: {
        "The": 1,
    },
    1: {
        "lazy": 1,
        "smelly": 1,
        "": 2
    },
    2: {
        "dog": 3,
        "cat": 3
    },
    3: {
        "ate": 4,
        "ran": 4
    },
    4: {
        "slowly": 5,
        "noisily": 5,
        "": 5
    },
    5: {
        "EOI": 6
    }
}


let validate_words = (words) => {
    curr_state = 0
    for (let word of words) {
        if (word in states_tab[curr_state]) {
            curr_state = states_tab[curr_state][word]
        }
        else if (word in states_tab[curr_state+1]) {
            curr_state = states_tab[curr_state+1][word]
        }
        else {
            return false
        }

        // END state
        if (curr_state == 6) {
            return true
        }
    }
    return false
}

// Testing
console.log("Begin testing")

const assert = require('assert').strict

assert(validate_words("The dog ran EOI".split(" ")))
assert(validate_words("The smelly lazy dog ran slowly EOI".split(" ")))
assert(validate_words("The lazy dog ran EOI".split(" ")))
assert(validate_words("The smelly cat ran EOI".split(" ")))
assert(validate_words("The smelly dog ate slowly EOI".split(" ")))
assert(validate_words("The lazy dog ate EOI".split(" ")))
assert(validate_words("The dog ran slowly EOI".split(" ")))
assert(validate_words("The dog ran".split(" ")) == false)
assert(validate_words("The ant ran EOI".split(" ")) == false)
assert(validate_words("The slowly ant ran EOI".split(" ")) == false)
assert(validate_words("The EOI".split(" ")) == false)
assert(validate_words("The lazy cat ate slowly".split(" ")) == false)

console.log("Done")